### PR TITLE
kvserver: add queue size metric to RaftTransport

### PR DESCRIPTION
**Pull request by pav-kv**
Currently, there is a `RaftEnqueuePending` metric in `StoreMetrics` which exposes
the `RaftTransport` outgoing queue size. However, this is a per-`Store` metric, so
it duplicates the same variable across all the stores. The metric should be
tracked on a per-node/transport basis.

This PR introduces a per-transport `SendQueueSize` metric to replace the old
`RaftEnqueuePending`. It also does the necessary plumbing to include it to
the exported metrics, since it's the first metric in `RaftTransport`.

Screenshot 2022-08-10 at 15 37 47

Touches #83917

Release note: None

### CODE REVIEW

**Comment by pavelkalinnikov** [comment]
@tbg PTAL

File: pkg/kv/kvserver/raft_transport.go [hunk 1]
```diff
@@ -301,6 +301,11 @@ func NewRaftTransport(
 	return t
 }
 
+// Metrics returns metrics tracking this transport.
+func (t *RaftTransport) Metrics() *RaftTransportMetrics {
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/raft_transport.go:305]: Alternatively, can do a `Register(r *metric.Register)` method, to avoid exposing the metrics.

**Review by tbg — COMMENTED** [review]
> Reviewed 4 of 4 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85885)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

___
*[`pkg/kv/kvserver/raft_transport.go` line 305 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85885#-N96uF1CZhG1ZHovwcGL-r2-305:-N97BK1vCqiFnJIs6REp:b9hhc34) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6e1b5bc34d902e77b27b9caaae099cdceb4fc91/pkg/kv/kvserver/raft_transport.go#L305)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Alternatively, can do a `Register(r *metric.Registry)` method, to avoid giving the caller access to the individual metrics.
</blockquote></details>

I think as-is is fine, since it's the pattern we use across the board.
___
*[`pkg/kv/kvserver/raft_transport_metrics.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85885#-N97BRkH8MhUbSaq4lsK:-N97BRkH8MhUbSaq4lsL:b-h3kad) ([raw file](https://github.com/cockroachdb/cockroach/blob/760d3d743f59640f778009da4f45fad9c3bf5397/pkg/kv/kvserver/raft_transport_metrics.go#L23)):*
> ```Go
> 	t.metrics = &RaftTransportMetrics{
> 		SendQueueSize: metric.NewFunctionalGauge(metric.Metadata{
> 			Name: "raft.transport.send-queue-size",
> ```

I was thinking we'd add this metric under the same name, to avoid churn and continue letting any Grafana dashboards that are out there ~probably continue to work.
I also would avoid the suffix "size" since it has a good chance to be misinterpreted as "size in bytes". Unfortunately we don't have the best naming conventions in place, so if you look at metrics names across the board you will notice that they're not principled.

File: pkg/kv/kvserver/raft_transport.go (hunk 1 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/raft_transport.go:305]: Leaving as is.

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85885)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/raft_transport_metrics.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85885#-N97BRkH8MhUbSaq4lsK:-N97D8rBAyKRWiGUV2UN:b-7khc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/760d3d743f59640f778009da4f45fad9c3bf5397/pkg/kv/kvserver/raft_transport_metrics.go#L23)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I was thinking we'd add this metric under the same name, to avoid churn and continue letting any Grafana dashboards that are out there ~probably continue to work.
I also would avoid the suffix "size" since it has a good chance to be misinterpreted as "size in bytes". Unfortunately we don't have the best naming conventions in place, so if you look at metrics names across the board you will notice that they're not principled.
</blockquote></details>

Technically, this metric still moves from per-store to per-node "keyspace" (see the screenshot in the PR description which shows the full metric path), so whoever relies on it currently can still find it missing if the sub-name is the same?

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85885)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/raft_transport_metrics.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85885#-N97BRkH8MhUbSaq4lsK:-N97FlJM2cfCnvEuQz4r:bpbjqau) ([raw file](https://github.com/cockroachdb/cockroach/blob/760d3d743f59640f778009da4f45fad9c3bf5397/pkg/kv/kvserver/raft_transport_metrics.go#L23)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Technically, this metric still moves from per-store to per-node "keyspace" (see the screenshot in the PR description which shows the full metric path), so whoever relies on it currently can still find it missing if the sub-name is the same?
</blockquote></details>

I anticipate having a byte size metric next to this one? How would you suggest to name them?

I was thinking of:
- "raft.transport.send-queue-size" for the size in messages
- "raft.transport.send-queue-bytes" for the size in bytes.

Also, the metric meta contains the unit `COUNT` and the "Messages", which for the bytes metric would be `BYTES` and "Bytes" correspondingly.

**Review by tbg — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85885)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

___
*[`pkg/kv/kvserver/raft_transport_metrics.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85885#-N97BRkH8MhUbSaq4lsK:-N97GVT-DFnz0rggtTBI:boikqwr) ([raw file](https://github.com/cockroachdb/cockroach/blob/760d3d743f59640f778009da4f45fad9c3bf5397/pkg/kv/kvserver/raft_transport_metrics.go#L23)):*
Sort of the only thing that matters is what it looks like to prometheus. The source node or store becomes a label. So if you re-use the old name, the metric goes from `foo{store=X}` to `foo{node=X}` and all your aggregation rules still work (as long as they didn't particularly depend on the `store`) label.

> How would you suggest to name them?

We don't have a convention so sort of anything goes. But looking at the metrics we have right now, I think we tend to name the "counts" without a suffix (i.s. like `raft.enqueued.pending`) and then give the corresponding bytes metric a `-bytes` suffix, for example here:

https://github.com/cockroachdb/cockroach/blob/fc85918139205d75e6de016e70fccd88237ce76f/pkg/kv/kvserver/metrics.go#L618-L629

I think `send-queue-{size,bytes}` is fine, and I don't think it matters much that we rename this metric, as there should be very little depending on it. So we can use the name you suggested.
___
*[`pkg/ui/workspaces/cluster-ui/src/store/nodes/nodes.fixtures.ts` line 809 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85885#-N97IA5K-GXgr5LGGO0T:-N97IA5K-GXgr5LGGO0U:bb0m9vk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c24f3564e213371bbe800e9646386cdefeb1546/pkg/ui/workspaces/cluster-ui/src/store/nodes/nodes.fixtures.ts#L809)):*
> ```XML
>           "queue.tsmaintenance.processingnanos": 174301000,
>           "raft.commandsapplied": 0,
>           "raft.enqueued.pending": 0,
> ```

Unclear to me if we also need to add new metrics here - I don't think we need to, but giving a courtesy tag to @cockroachdb/kv-obs-prs as they'll hopefully know if something needs to be done. (No need to wait if CI passes, I think).

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85885)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

___
*[`pkg/kv/kvserver/raft_transport_metrics.go` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85885#-N97KcdV6rZUiq1_6_cc:-N97KcdV6rZUiq1_6_cd:bsql2ov) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6e1b5bc34d902e77b27b9caaae099cdceb4fc91/pkg/kv/kvserver/raft_transport_metrics.go#L26)):*
> ```Go
> 			Help: `Number of pending outgoing messages in the Raft Transport queue.
> 
> The queue is bounded in size, so instead of unbounded growth one would observe a
> ```

BTW, this seems not true. This metric accumulates all per-peer channel sizes, so is limited by `totalNodes * 10k`. It can be hard to analyse. E.g. if this metric is 10k on a cluster of 3 nodes, it can be ok (if the load is uniform, i.e. 5k per peer), or bad (all 10k backlog is on one particular peer).



_Code quote:_
```Go
The queue is bounded in size, so instead of unbounded growth one would observe a
ceiling value in the tens of thousands.`,
```

**Review by tbg — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85885)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

___
*[`pkg/kv/kvserver/raft_transport_metrics.go` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85885#-N97KcdV6rZUiq1_6_cc:-N97R0W11DYyeFq6TLa5:b-t4yo9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6e1b5bc34d902e77b27b9caaae099cdceb4fc91/pkg/kv/kvserver/raft_transport_metrics.go#L26)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

BTW, this seems not true. This metric accumulates all per-peer channel sizes, so is limited by `totalNodes * 10k`. It can be hard to analyse. E.g. if this metric is 10k on a cluster of 3 nodes, it can be ok (if the load is uniform, i.e. 5k per peer), or bad (all 10k backlog is on one particular peer).
</blockquote></details>

Good point, everything generally over-indexes on the one store one node case, but this is not always the case. (We have customers running with 8-12 stores per node). Mind updating the verbiage here, maybe saying that values in the thousands or even tens of thousands could indicate that there are issues streaming messages to at least one peer.

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85885)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov and @tbg)

___
*[`pkg/kv/kvserver/raft_transport_metrics.go` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85885#-N97KcdV6rZUiq1_6_cc:-N97a6Js7TTarjUy4lPI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6e1b5bc34d902e77b27b9caaae099cdceb4fc91/pkg/kv/kvserver/raft_transport_metrics.go#L26)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Good point, everything generally over-indexes on the one store one node case, but this is not always the case. (We have customers running with 8-12 stores per node). Mind updating the verbiage here, maybe saying that values in the thousands or even tens of thousands could indicate that there are issues streaming messages to at least one peer.
</blockquote></details>

Done.

**Comment by pavelkalinnikov** [comment]
bors r=tbg

**Pull request merged**